File: open-mic-server.py
#!/usr/bin/env python3
"""
open-mic-server.py - PC web service for OpenCodeMic

Listens for POST requests from the OpenCodeMic Android app and sends
keystrokes to the active window and/or the opencode GUI via CDP bridge.

Usage:
    python3 open-mic-server.py
    python3 open-mic-server.py --password mysecret
    python3 open-mic-server.py --password mysecret --https
    python3 open-mic-server.py --https
"""
import argparse, json, os, re, subprocess, sys, threading, ssl
from http.server import HTTPServer, BaseHTTPRequestHandler
import cdp_bridge
import scheduler
import conversation_logger

BUFFER_FILE = '/tmp/opencode_mic_buffer.txt'


class ServerState:
    def __init__(self):
        self.focus_mode = True
        self.paused = False
        self.auto_exec = False
        self.auto_exec_timer = None
        self.last_sent = ''
        self.word_counts = []


state = ServerState()


def read_buffer():
    if not os.path.exists(BUFFER_FILE):
        return ''
    try:
        with open(BUFFER_FILE) as f:
            return f.read()
    except OSError:
        return ''


def save_buffer(content):
    try:
        with open(BUFFER_FILE, 'w') as f:
            f.write(content)
    except OSError as e:
        print(f"save_buffer: {e}", file=sys.stderr)


def clear_buffer():
    try:
        os.unlink(BUFFER_FILE)
    except FileNotFoundError:
        pass


def send_cdp_keys(*keys):
    if not state.focus_mode:
        return
    key_str = ','.join(keys)
    try:
        if key_str == 'Escape':
            cdp_bridge.do_escape(3)
        elif key_str == 'Enter':
            cdp_bridge.do_enter()
        elif key_str == 'Tab':
            cdp_bridge.do_tab()
        elif key_str == 'C-w':
            cdp_bridge.do_delete_word()
        elif key_str == 'C-u':
            cdp_bridge.do_clear_line()
    except Exception as e:
        print(f"CDP ERROR: {e} — is the opencode GUI running?")


def send_text(text):
    if state.focus_mode:
        try:
            cdp_bridge.do_type_text(text)
        except Exception as e:
            print(f"CDP ERROR: {e} — is the opencode GUI running?")
    else:
        subprocess.run(['xdotool', 'type', text])


def cancel_auto_exec():
    if state.auto_exec_timer:
        state.auto_exec_timer.cancel()
        state.auto_exec_timer = None


def spawn_auto_exec():
    if not state.auto_exec:
        return
    cancel_auto_exec()
    timer = threading.Timer(2.0, do_auto_enter)
    timer.daemon = True
    timer.start()
    state.auto_exec_timer = timer


def do_auto_enter():
    if state.focus_mode:
        cdp_bridge.do_enter()
    else:
        subprocess.run(['xdotool', 'key', 'Enter'])


BEEP_FILE = '/home/rj/su/computerbeep_69.mp3'


def process_text(chunk):
    chunk = re.sub(r'\[.+?\]', '', chunk)
    chunk = chunk.strip()
    if not chunk:
        return

    # "computer" prefix — play beep; if just "computer" alone, drop it
    if chunk.lower().startswith('computer'):
        try:
            subprocess.Popen(
                ['mpg123', '-q', BEEP_FILE],
                stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
            )
        except FileNotFoundError:
            try:
                subprocess.Popen(
                    ['ffplay', '-nodisp', '-autoexit', '-loglevel', 'quiet', BEEP_FILE],
                    stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
                )
            except FileNotFoundError:
                print("WARNING: no mpg123 or ffplay found for beep")
        # If chunk is exactly "computer" (or "computer "), drop it
        if chunk.lower().strip() == 'computer':
            return

    buffer = read_buffer()
    combined = (buffer + " " + chunk).strip()

    if re.search(r'\bstop\W*stop\b', combined, re.IGNORECASE):
        print("KEYWORD: stop stop -> Escape x3")
        for _ in range(3):
            send_cdp_keys('Escape')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bgo\W*go\b', combined, re.IGNORECASE):
        print("KEYWORD: go go -> Enter")
        send_cdp_keys('Enter')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\b(?:enter|execute)\b', combined, re.IGNORECASE):
        print("KEYWORD: enter/execute -> Enter")
        send_cdp_keys('Enter')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bundo\b', combined, re.IGNORECASE):
        n = len(re.findall(r'\bundo\b', combined, re.IGNORECASE))
        print(f"KEYWORD: undo x {n}")
        for _ in range(n):
            if not state.word_counts:
                break
            wc = state.word_counts.pop()
            for _ in range(wc):
                send_cdp_keys('C-w')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bdelete\W*word\b', combined, re.IGNORECASE):
        print("KEYWORD: delete word -> C-w")
        send_cdp_keys('C-w')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\b(?:clear\W*line|erase\W*text)\b', combined, re.IGNORECASE):
        print("KEYWORD: clear line/erase text -> C-u")
        send_cdp_keys('C-u')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\btab\b|\bnext\W*agent\b', combined, re.IGNORECASE):
        print("KEYWORD: tab/next agent -> Tab")
        send_cdp_keys('Tab')
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bpause\W*work\b', combined, re.IGNORECASE):
        state.paused = True
        print(f"KEYWORD: pause work -> paused={state.paused}")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bresume\W*work\b', combined, re.IGNORECASE):
        state.paused = False
        print(f"KEYWORD: resume work -> paused={state.paused}")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bfocus\W*off\b', combined, re.IGNORECASE):
        state.focus_mode = False
        print(f"KEYWORD: focus off -> focus_mode={state.focus_mode}")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bfocus\W*on\b', combined, re.IGNORECASE):
        state.focus_mode = True
        print(f"KEYWORD: focus on -> focus_mode={state.focus_mode}")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\b(?:focus|activate)\W*terminal\b', combined, re.IGNORECASE):
        print("KEYWORD: focus/activate terminal -> GUI")
        try:
            result = subprocess.run(
                ['xdotool', 'search', '--name', '^OpenCode$'],
                capture_output=True, text=True, timeout=5
            )
            wid = result.stdout.strip()
            if wid:
                subprocess.run(['wmctrl', '-i', '-a', wid], timeout=5)
        except (subprocess.TimeoutExpired, FileNotFoundError):
            pass
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\benable\W*automatic\W*execution\b', combined, re.IGNORECASE):
        state.auto_exec = True
        print("KEYWORD: enable automatic execution")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r'\bdisable\W*automatic\W*execution\b', combined, re.IGNORECASE):
        state.auto_exec = False
        cancel_auto_exec()
        print("KEYWORD: disable automatic execution")
        clear_buffer()
        state.last_sent = ''
        return

    if re.search(r"^i['´`]?m\s+not\s+sure\.?\s*$", chunk, re.IGNORECASE):
        print(f"FILTER (hallucination): {chunk}")
        return
    if re.search(r"^i['´`]?m\s+not\s+going\s+to\s+get\s+it\.?\s*$", chunk, re.IGNORECASE):
        print(f"FILTER (hallucination): {chunk}")
        return
    if re.search(r'\bclick\b', chunk, re.IGNORECASE):
        print(f"FILTER (click): {chunk}")
        return

    if state.paused:
        print(f"PAUSED (dropped): {chunk}")
        return

    if chunk == state.last_sent:
        print(f"SKIP (duplicate): {chunk}")
        return

    chunk = re.sub(r'\bperiod\b', '.', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bcomma\b', ',', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bquestion\s*mark\b', '?', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bexclamation\s*mark\b', '!', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bdash\b', '-', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bslash\b', '/', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bcolon\b', ':', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r'\bsemicolon\b', ';', chunk, flags=re.IGNORECASE)
    chunk = re.sub(r' ([.,!?:;])', r'\1', chunk)

    to_send = f" {chunk}"
    print(f"SEND:{to_send}", end='')
    send_text(to_send)
    print(" [OK]")
    state.last_sent = chunk

    wc = len(chunk.split())
    if wc > 0:
        state.word_counts.append(wc)

    tail = chunk[-10:] if len(chunk) > 10 else chunk
    save_buffer(tail)
    spawn_auto_exec()


class Handler(BaseHTTPRequestHandler):
    password = ''

    def do_POST(self):
        if self.password:
            auth = self.headers.get('Authorization', '')
            if auth != f'Bearer {self.password}':
                self.send_response(401)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'unauthorized'}).encode())
                return
        length = int(self.headers.get('Content-Length', 0))
        raw = self.rfile.read(length).decode()
        if raw:
            try:
                data = json.loads(raw)
                text = data.get('text', '')
                if text:
                    process_text(text)
            except json.JSONDecodeError:
                pass
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps({'ok': True}).encode())

    def log_message(self, *args):
        pass


def ensure_cert(cert_path, key_path):
    """Generate a self-signed cert if missing."""
    if os.path.exists(cert_path) and os.path.exists(key_path):
        return
    import shutil
    if not shutil.which('openssl'):
        print("WARNING: openssl not found. Generate a certificate manually:")
        print(f"  openssl req -x509 -newkey rsa:2048 -keyout {key_path} -out {cert_path} -days 365 -nodes -subj /CN=OpenCodeMic")
        print("Running without HTTPS...", file=sys.stderr)
        return
    subprocess.run([
        'openssl', 'req', '-x509', '-newkey', 'rsa:2048',
        '-keyout', key_path, '-out', cert_path,
        '-days', '365', '-nodes',
        '-subj', '/CN=OpenCodeMic'
    ], capture_output=True)
    if os.path.exists(cert_path) and os.path.exists(key_path):
        print(f"Generated self-signed cert: {cert_path}, {key_path}")
    else:
        print("WARNING: failed to generate cert. Running without HTTPS...", file=sys.stderr)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='OpenCodeMic server')
    parser.add_argument('port', nargs='?', type=int, default=9876, help='Port to listen on')
    parser.add_argument('--https', action='store_true', help='Enable HTTPS with auto-generated self-signed cert')
    parser.add_argument('--cert', help='SSL certificate file (default: ~/.config/opencode-mic/cert.pem)')
    parser.add_argument('--key', help='SSL key file (default: ~/.config/opencode-mic/key.pem)')
    parser.add_argument('--password', help='Shared secret for client authentication')
    args = parser.parse_args()

    Handler.password = args.password or ''

    if args.password and not args.https and not (args.cert and args.key):
        print(f"WARNING: password set without HTTPS — transmitted in plaintext!")

    if args.https:
        cert_dir = os.path.expanduser('~/.config/opencode-mic')
        os.makedirs(cert_dir, exist_ok=True)
        args.cert = args.cert or os.path.join(cert_dir, 'cert.pem')
        args.key = args.key or os.path.join(cert_dir, 'key.pem')

    if args.cert and args.key:
        ensure_cert(args.cert, args.key)
        if os.path.exists(args.cert) and os.path.exists(args.key):
            context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
            context.load_cert_chain(args.cert, args.key)
            server = HTTPServer(('0.0.0.0', args.port), Handler)
            server.socket = context.wrap_socket(server.socket, server_side=True)
            proto = 'https'
            tag = ' + HTTPS' if args.password else ''
            print(f"Running with HTTPS{tag} on port {args.port}")
        else:
            print("WARNING: cert generation failed, falling back to HTTP")
            server = HTTPServer(('0.0.0.0', args.port), Handler)
            proto = 'http'
    else:
        server = HTTPServer(('0.0.0.0', args.port), Handler)
        proto = 'http'

    logger = conversation_logger.ConversationLogger()
    logger_thread = threading.Thread(
        target=conversation_logger.poll_loop, args=(logger,),
        daemon=True, name='logger'
    )
    logger_thread.start()
    scheduler.start(logger)
    print(f"OpenCodeMic server listening on {proto}://0.0.0.0:{args.port}")
    print(f"Buffer: {BUFFER_FILE}")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down")
        server.server_close()

File: open-mic-server.pl
#!/usr/bin/env perl
use strict;
use warnings;
use JSON::PP;
use HTTP::Server::PSGI;

my $buffer_file = '/tmp/opencode_mic_buffer.txt';
my $last_sent = '';
my $paused = 0;
my $focus_mode = 1;
my $auto_exec = 0;
my $auto_exec_pid = 0;
my @word_counts = ();

my $cdp_bridge = '/home/rj/su/cdp_bridge.py';
$SIG{CHLD} = 'IGNORE';

sub send_keys {
    my (@keys) = @_;
    return unless $focus_mode;
    my $key_str = join(',', @keys);
    if ($key_str eq 'Escape') {
        system("python3 $cdp_bridge escape 3 >/dev/null 2>&1");
    } elsif ($key_str eq 'Enter') {
        system("python3 $cdp_bridge enter >/dev/null 2>&1");
    } elsif ($key_str eq 'Tab') {
        system("python3 $cdp_bridge agent_next >/dev/null 2>&1");
    } elsif ($key_str eq 'C-w') {
        system("python3 $cdp_bridge delete_word >/dev/null 2>&1");
    } elsif ($key_str eq 'C-u') {
        system("python3 $cdp_bridge clear_line >/dev/null 2>&1");
    }
}

sub send_text {
    my ($text) = @_;
    if ($focus_mode) {
        my $escaped = $text;
        $escaped =~ s/'/'\\''/g;
        system("python3 $cdp_bridge text '$escaped' >/dev/null 2>&1");
    } else {
        system('xdotool', 'type', $text);
    }
}

sub spawn_auto_exec_timer {
    return unless $auto_exec;
    if ($auto_exec_pid && kill 0, $auto_exec_pid) {
        kill 'TERM', $auto_exec_pid;
    }
    $auto_exec_pid = fork;
    if ($auto_exec_pid == 0) {
        close STDIN; close STDOUT; close STDERR;
        sleep 2;
        send_keys('Enter');
        exit 0;
    }
}

sub cancel_auto_exec_timer {
    if ($auto_exec_pid && kill 0, $auto_exec_pid) {
        kill 'TERM', $auto_exec_pid;
    }
    $auto_exec_pid = 0;
}

sub app {
    my $env = shift;

    if ($env->{REQUEST_METHOD} eq 'POST' && $env->{PATH_INFO} eq '/') {
        my $content;
        my $input = $env->{'psgi.input'};
        read($input, $content, $env->{CONTENT_LENGTH} || 0);

        if ($content) {
            my $data = JSON::PP::decode_json($content);
            my $text = $data->{text} // '';
            process_text($text) if length $text;
        }

        return [200, ['Content-Type' => 'application/json'], [JSON::PP::encode_json({ok => JSON::PP::true})]];
    }

    return [404, ['Content-Type' => 'application/json'], [JSON::PP::encode_json({error => 'not found'})]];
}

sub process_text {
    my ($chunk) = @_;

    $chunk =~ s/\[.+?\]//g;
    $chunk =~ s/^\s+|\s+$//g;
    return if $chunk eq '';

    my $buffer = '';
    if (-f $buffer_file) {
        open my $fh, '<', $buffer_file or warn "read_buffer: $!";
        if ($fh) {
            $buffer = <$fh>;
            close $fh;
        }
    }

    my $combined = $buffer . " " . $chunk;

    if ($combined =~ /\bstop\W*stop\b/i) {
        print "KEYWORD: stop stop -> Escape x3\n";
        for (1 .. 3) {
            send_keys('Escape');
        }
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bgo\W*go\b/i) {
        print "KEYWORD: go go -> Enter\n";
        send_keys('Enter');
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\b(?:enter|execute)\b/i) {
        print "KEYWORD: enter/execute -> Enter\n";
        send_keys('Enter');
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\b(?:undo|revert)\b/i) {
        my $n = () = $combined =~ /\b(?:undo|revert)\b/gi;
        print "KEYWORD: undo/revert x $n\n";
        for (1 .. $n) {
            my $wc = pop @word_counts or last;
            send_keys(('C-w') x $wc);
        }
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bdelete\W*word\b/i) {
        print "KEYWORD: delete word -> C-w\n";
        send_keys('C-w');
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\b(?:clear\W*line|erase\W*text)\b/i) {
        print "KEYWORD: clear line/erase text -> C-u\n";
        send_keys('C-u');
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\btab\b|\bnext\W*agent\b/i) {
        print "KEYWORD: tab/next agent -> Tab\n";
        send_keys('Tab');
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bpause\W*work\b/i) {
        $paused = 1;
        print "KEYWORD: pause work -> paused=$paused\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bresume\W*work\b/i) {
        $paused = 0;
        print "KEYWORD: resume work -> paused=$paused\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bfocus\W*off\b/i) {
        $focus_mode = 0;
        print "KEYWORD: focus off -> focus_mode=$focus_mode\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bfocus\W*on\b/i) {
        $focus_mode = 1;
        print "KEYWORD: focus on -> focus_mode=$focus_mode\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\b(?:focus|activate)\W*terminal\b/i) {
        print "KEYWORD: focus/activate terminal -> GUI\n";
        my $wid = `xdotool search --name "^OpenCode$"`;
        chomp $wid;
        system('wmctrl', '-i', '-a', $wid) if $wid;
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\benable\W*automatic\W*execution\b/i) {
        $auto_exec = 1;
        print "KEYWORD: enable automatic execution\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    if ($combined =~ /\bdisable\W*automatic\W*execution\b/i) {
        $auto_exec = 0;
        cancel_auto_exec_timer();
        print "KEYWORD: disable automatic execution\n";
        clear_buffer();
        $last_sent = '';
        return;
    }

    # filter known hallucinations
    if ($chunk =~ /^i['´`]?m\s+not\s+sure\.?\s*$/i) {
        print "FILTER (hallucination): $chunk\n";
        return;
    }
    if ($chunk =~ /^i['´`]?m\s+not\s+going\s+to\s+get\s+it\.?\s*$/i) {
        print "FILTER (hallucination): $chunk\n";
        return;
    }
    if ($chunk =~ /click/i) {
        print "FILTER (click): $chunk\n";
        return;
    }

    # if paused, discard non-keyword chunks
    if ($paused) {
        print "PAUSED (dropped): $chunk\n";
        return;
    }

    # dedup: skip if same as last chunk
    if ($chunk eq $last_sent) {
        print "SKIP (duplicate): $chunk\n";
        return;
    }

    # replace punctuation words with symbols
    $chunk =~ s/\bperiod\b/./gi;
    $chunk =~ s/\bcomma\b/,/gi;
    $chunk =~ s/\bquestion\s*mark\b/?/gi;
    $chunk =~ s/\bexclamation\s*mark\b/!/gi;
    $chunk =~ s/\bdash\b/-/gi;
    $chunk =~ s/\bslash\b/\//gi;
    $chunk =~ s/\bcolon\b/:/gi;
    $chunk =~ s/\bsemicolon\b/;/gi;
    $chunk =~ s/ ([.,!?:;])/$1/g;

    my $to_send = " $chunk";
    print "SEND:$to_send";
    send_text($to_send);
    print " [OK]\n";
    $last_sent = $chunk;

    my $wc = scalar(split(/\s+/, $chunk));
    push @word_counts, $wc if $wc > 0;

    my $tail = length($chunk) > 10 ? substr($chunk, -10) : $chunk;
    save_buffer($tail);
    spawn_auto_exec_timer();
}

sub clear_buffer {
    unlink $buffer_file;
}

sub save_buffer {
    my ($content) = @_;
    open my $fh, '>', $buffer_file or warn "save_buffer: $!";
    if ($fh) {
        print $fh $content;
        close $fh;
    }
}

unless (caller) {
    my $port = shift(@ARGV) || 9876;
    my $server = HTTP::Server::PSGI->new(
        host => '0.0.0.0',
        port => $port,
    );
    print "OpenCodeMic server listening on http://0.0.0.0:$port\n";
    print "Buffer: $buffer_file\n";
    $server->run(\&app);
}

1;
__END__
=head1 NAME

open-mic-server.pl - PC web service for OpenCodeMic

=head1 SYNOPSIS

    perl open-mic-server.pl [port]

    # or with plackup:
    plackup open-mic-server.pl

=head1 DESCRIPTION

Listens for POST requests from the OpenCodeMic Android app.
Sends keystrokes to the active window and the opencode GUI via CDP.

Modes:
    "focus on"   - Send to opencode GUI via CDP bridge (default)
    "focus off"  - xdotool type to the active window only (no CDP, no key combos)
    "activate terminal" - Bring the opencode GUI window to the foreground

Keywords:    "stop stop"  - Send Escape x3 (cancel/clear)
    "go go"      - Send Enter (submit)
    everything else - Type the text
